doc/en/fixture.rst

```diff
@@ -115,48 +121,521 @@ for reference:
+much everything except for the "**act**". It's lining up the dominoes so that
+the **act** can do its thing in one, state-changing step. This can mean
+preparing objects, starting/killing services, enterring records into a database,
+or even things like defining a URL to query, generating some credentials for a
+user that doesn't exist yet, or just waiting for some process to finish.
+
+**Act** is the singular, state-changing action that kicks off the **behavior**
+we want to test. This behavior is what carries out the changing of the SUT's
+state, and it's the resulting changed state that we can look at to make a
+judgement about the behavior. This typically takes the form of a function/method
+call.
+
+**Assert** is where we look at that resulting state and check if it looks how
+we'd expect after the dust has settled. It's where we gather evidence to say the
+behavior does or does not aligns with what we expect. The ``assert`` in our test
+is where we take that measurement/observation and apply our judgement to it. If
+something should be green, we'd say ``assert thing == "green"``.
+
+**Cleanup** is where the test picks up after itself, so other tests aren't being
+accidentally influenced by it.
 
-Test functions can receive fixture objects by naming them as an input
-argument. For each argument name, a fixture function with that name provides
-the fixture object.  Fixture functions are registered by marking them with
-:py:func:`@pytest.fixture <_pytest.python.fixture>`.  Let's look at a simple
-self-contained test module containing a fixture and a test function
-using it:
+At it's core, the test is ultimately the **act** and **assert** steps, with the
+**arrange** step only providing the context. **Behavior** exists between **act**
+and **assert**.
+
+.. note::
+
+    For functional programming, the **act** will still be a function call, but
+    the **assert** will be looking at the output of that function, rather than
+    the resulting state. But fixtures can still be used for this if desired.
+
+Back to fixtures
+^^^^^^^^^^^^^^^^
+
+"Fixtures", in the literal sense, are each of the **arrange** steps and data. They're
+everything that test needs to do it's thing.
+
+In pytest, "fixtures" are functions you define that serve this purpose. But they
+don't have to be limited to just the **arrange** steps. They can provide the
+**act** step, as well, and this can be a powerful technique for designing more
+complex tests, especially given how pytest's fixture system works. But we'll get
+into that further down.
+
+We can tell pytest that a particular function is a fixture by deocrating it with
+:py:func:`@pytest.fixture <_pytest.python.fixture>`. Here's a simple example of
+what a fixture in pytest might look like:
 
 .. code-block:: python
 
-    # content of ./test_smtpsimple.py
     import pytest
 
 
     @pytest.fixture
-    def smtp_connection():
-        import smtplib
+    def order():
```

Comment:
I agree, having some recognizable meaning attached can help make it clear what should happen when. However, I think the SMTP stuff might throw off some folks (e.g. I knew what SMTP was, but didn't know about `HELO`/`EHLO`). What about something like this?

```python
 import pytest

from google import GmailClient  # Note: not a real library


@pytest.fixture
def login_response():
    return GmailClient("myusername", "mypassword")

def test_login_response_status_code(login_response):
    # status code of 200 means a successful login
    assert login_response.status_code == 200
```

Reply:
> What about something like this?

Seems good, the problem however is that we actually *run* those examples to generate correct output (using `tox -e regen`), so those examples need to be executable.

I think we can change the examples to use something:

1. Is from the standard-library.
2. Runs in-memory, without needing to change things on disk.
3. Doesn't need servers or other external resources.

Some thoughts come to mind: testing `json` encoding/decoding, testing something from the `collections` module (`Counter` for example), `datetime`, `fractions`, etc.

Reply:
@nicoddemus @bluetech ok, I updated it a bit, but haven't fleshed it out for the rest of the doc, and wanted to get your feedback. I couldn't think of any examples that both could be relatable and also wouldn't run the risk of making a readers eyes glaze over, so I made some `Fruit` and `FruitSalad` examples. I've been thinking that food related stuff could be used in the examples, because cooking involves branching paths and sequential steps (and even some async stuff if we want to go that route), so it seemed fitting. Let me know what you think!

Reply:
Thanks @SalmonMode, I think those are OK to read and understand.